File: datanode/rdfdia.rb
#!/usr/bin/env ruby


require 'rubygems'
require 'set'
require 'trollop'
require 'rdf'
require 'cgi'
require 'rdf/raptor'
require 'yaml'
load 'gentree.rb'

class Dia

  
	def initialize(file, schema, modifier)
		@schema = schema
    @labelStyle = 'prefixed' # or 'mixed'
    @ress = Set.new
		@tree = {}
    @modifier = modifier
		RDF::Reader.open(file) do |reader|
			reader.each_statement do |st|
				if !@tree.include? st.subject.to_s
					@tree[st.subject.to_s] = {}
				end
				if !@tree[st.subject.to_s].include? st.predicate.to_s
					@tree[st.subject.to_s][st.predicate.to_s] = []
				end
				@tree[st.subject.to_s][st.predicate.to_s].push st.object.to_s
        @ress.add st.subject.to_s
        @ress.add st.predicate.to_s
        if st.object.kind_of?(RDF::Resource)
          @ress.add st.object.to_s
        end
			end
		end
	end
	
  def labelStyle
    @labelStyle || 'prefixed'
  end
  
  def prefixed(o)
		prefix = {
			"http://yago-knowledge.org/resource/" => "yago",
			"http://www.w3.org/1999/02/22-rdf-syntax-ns#" => "rdf",
			"http://xmlns.com/foaf/0.1/" => "foaf",
			"http://dbpedia.org/property/" => "dbp",
			"http://www.w3.org/2002/07/owl#" => "owl",
			"http://purl.org/dc/elements/1.1/" => "dc",
			"http://www.w3.org/2000/01/rdf-schema#" => "rdfs",
			"http://dbpedia.org/ontology/" => "dbo",
			"http://purl.org/rss/1.0/" => "rss",
			"http://www.w3.org/2004/02/skos/core#" => "skos",
			"http://rdfs.org/ns/void#" => "void",
			"http://purl.org/datanode/ns/" => "dn",
			"http://purl.org/datanode/ex/" => "ex",
			"http://purl.org/dc/terms/" => "dct",
		}
		prefix.each do |ns, pref|
			if o.start_with?(ns) then
				localp = ""
				if o.include?("#") then
					localp = o[o.index("#")+1, o.length]
				else
					localp = o[o.rindex("/")+1, o.length]
				end
				return "#{pref}:#{localp}"
			end
		end
		return o
  end
  
  def localname(o)
		if o.include?("#") then
      return o[o.index("#")+1, o.length]
    else 
      if o.include?("/") then
			  return o[o.rindex("/")+1, o.length]
      else
        return o
      end
		end
  end
  
	def label(o, useHtml = false)
		if !@schema.tree.include?(o) then 
			if o.start_with? "http" then
				return prefixed o
      else
        if o.start_with? "_:" then
          return o.to_s
        else
				  # This is a value
          if useHtml == true then
				    html = o.gsub(/([^\.?!]{20,30}[\.?!])|([^,0-9]{20,30},)/, '\0<br/>').gsub(/[0-9]{1,3}\)/, '<br/>\0')
            html = CGI.escapeHTML(html);
            html = "<html>#{html}</html>"
				    html = html.gsub(/\</,'&lt;').gsub(/\>/,'&gt;')
				    return html
          else
            return o
          end
        end
			end
		end
		vals = @schema.tree[o]
    if labelStyle == 'prefixed' then return prefixed(o) end
    # take the label from the vocabulary 
		return vals.include?("http://www.w3.org/2000/01/rdf-schema#label") ? vals["http://www.w3.org/2000/01/rdf-schema#label"].first : nil
	end
	
	def entities
    #ents = Set.new
    ents = Array.new
		@tree.each do |s,ps|
      # we skip foaf:Document
		  if @tree[s]['http://www.w3.org/1999/02/22-rdf-syntax-ns#type'] && @tree[s]['http://www.w3.org/1999/02/22-rdf-syntax-ns#type'].include?('http://xmlns.com/foaf/0.1/Document') then
        # do nothing
      else
			  #ents.add s
        if !(ents.include? s) then
          ents.push s  
        end
			  ps.each do |p, vals|
				  vals.each do |v| 
            if !(ents.include? v) then
					    if @modifier == 'all' or @ress.include? v then
                ents.push v
              end
            end
				  end
			  end
      end
		end
		return ents
	end
	
  def graphml()
			puts "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>"
			puts "<graphml \n\txmlns=\"http://graphml.graphdrawing.org/xmlns\" "
			puts "\txmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" "
			puts "\txmlns:y=\"http://www.yworks.com/xml/graphml\" "
			puts "\txmlns:yed=\"http://www.yworks.com/xml/yed/3\" "
			puts "\txsi:schemaLocation=\"http://graphml.graphdrawing.org/xmlns http://www.yworks.com/xml/schema/graphml/1.1/ygraphml.xsd\">"
			puts "\t\t<key attr.name=\"label\" attr.type=\"string\" for=\"node\" id=\"k_label\"/>"
			puts "\t\t<key for=\"node\" id=\"key_nodegraphics\" yfiles.type=\"nodegraphics\"/>"
			puts "\t\t<key for=\"edge\" id=\"key_edgegraphics\" yfiles.type=\"edgegraphics\"/>"
			#puts "<key attr.name=\"description\" attr.type=\"string\" for=\"node\" id=\"d5\"/>"
			
			puts "\t<graph edgedefault=\"directed\" id=\"G\">"
			
			entities.each do |node|
				puts "<!-- #{node} -->"
				nodel = label(localname(node), true)
				puts "<!-- #{nodel} -->"
				puts "\t\t<node id=\"n#{node.hash}\">"
				puts "\t\t\t<data key=\"k_label\"><![CDATA[#{node}]]></data>"
				puts "\t\t\t<data key=\"d5\"><![CDATA[#{node}]]></data>"
				puts "\t\t\t<data key=\"key_nodegraphics\">"
        		puts "\t\t\t\t<y:GenericNode configuration=\"com.yworks.flowchart.start1\">"
          		puts "\t\t\t\t\t<y:Fill color=\"#FFFFFF\" transparent=\"true\"/>"
          		puts "\t\t\t\t\t<y:BorderStyle color=\"#FFFFFF\" transparent=\"true\" type=\"line\" width=\"0.0\"/>"
          		puts "\t\t\t\t\t<y:NodeLabel "
				print ">#{nodel}<y:LabelModel><y:SmartNodeLabelModel distance=\"4.0\"/></y:LabelModel> "
				print "<y:ModelParameter> "
				print "<y:SmartNodeLabelModelParameter "
				print "/>"
				print "</y:ModelParameter>"
				print "</y:NodeLabel>"
        		puts "\t\t\t\t</y:GenericNode>"
				puts "\t\t\t</data>"
				puts "\t\t</node>"
			end
			@tree.each do |node, links|
				if entities.include? node then
				  edgeId=0
				  # parents
				  links.each do |link, to|
					  to.each do |o|
					    puts "<!-- #{to} -->"
						  edgeId = node.hash.to_s + "_" + link.hash.to_s + "_" + o.hash.to_s
						  edgel = label link, true
						  puts "\t\t<edge id=\"e_#{edgeId}\" source=\"n#{node.hash}\" target=\"n#{o.hash}\">"
						  puts "\t\t\t<data key=\"key_edgegraphics\">"
				      puts "\t\t\t<y:BezierEdge>"
						  puts "\t\t\t\t<y:LineStyle color=\"#000000\" type=\"line\" width=\"2.0\"/>"
              puts "\t\t\t\t<y:Arrows source=\"none\" target=\"white_delta\"/>"
					    puts "\t\t\t\t<y:EdgeLabel>#{edgel}</y:EdgeLabel>"
						  puts "\t\t\t</y:BezierEdge>"
						  puts "\t\t\t</data>"
						  puts "\t\t</edge>"
					  end
				  end
        end
			end
			
			puts "\t</graph>"
			puts "</graphml>"
	end
  
  def jsembed(o)
    return o.to_s.gsub("\"", "\\\"")
  end
  
  def jsobject()
    print "["
    ents = entities
    for index in 0...ents.size 
      node=ents[index]
      puts "{ "
      ix=index
      lbl=label node
      puts "\tid: \"#{jsembed ix}\","
      puts "\turi: \"#{jsembed node}\","
      puts "\tlabel: \"#{jsembed lbl}\","
      print "\tlinks: ["
			@tree.each do |n, links|
				if n == node then
			  links.each do |link, to|
			 to.each do |o|
         if ents.include? o then
            puts "{"
            tg=ents.index o
            puts "\t\ttarget: #{tg},"
					  edgel = label link
            puts "\t\tlabel: \"#{edgel}\","
            puts "\turi: \"#{link.to_s}\","
            puts "\t\ttype: \"default\","
            puts "\t\tweight: 1,"
            print "\t},"
          end
          end
      end
    end
  end
      puts "],"
      print "},"
  end
    puts "]"
  end
	
end

if __FILE__ == $0

	opts = Trollop::options do
    version "Voctoys 1.0.0 (c) 2013 Enrico Daga"
	  opt :file, "RDF file", :short => "-f", :type => :string
	  opt :voc, "RDF vocabulary", :short => "-v", :type => :string
	  opt :output, "Output format: graphml|js", :short => "-o", :type => :string
	  opt :modifier, "Modifier: links|all", :short => "-m", :type => :string, :default => "all"
	end
	
	Trollop::die :file, "must exist" unless File.exist?(opts[:file]) if opts[:file]

  $stdout.sync = true

	file = opts[:file]
	voc = opts[:voc]
  output = opts[:output]
  modifier = opts[:modifier]
	vok = Vok.new voc
	dia = Dia.new file, vok, modifier
  if output == 'graphml' then
    dia.graphml
  end
  if output == 'js' then
    puts "var nodes="
    dia.jsobject
  end
end

File: datanode/rdfdia.py
#!/usr/bin/env python
# 

import sys
import os.path
import os
import json
import argparse
import re
from rdflib import Graph, Literal, BNode, Namespace, RDF, URIRef
from rdflib.namespace import RDF, OWL, RDFS, NamespaceManager

from rdftree import Vok

import logging
logging.basicConfig()

class Dia:

    def __init__(self, data, vok, modifier='all'):
        # body of the constructor
        self._datafile = data
        g = Graph()
        self._data = g.parse(self._datafile, format="turtle")
        self._namespaces = NamespaceManager(self._data)
        self._vok = vok
        self._modifier = modifier
        self._namespaces.bind("yago","http://yago-knowledge.org/resource/")
        self._namespaces.bind("rdf","http://www.w3.org/1999/02/22-rdf-syntax-ns#")
        self._namespaces.bind("foaf","http://xmlns.com/foaf/0.1/")
        self._namespaces.bind("dbp","http://dbpedia.org/property/")
        self._namespaces.bind("owl","http://www.w3.org/2002/07/owl#")
        self._namespaces.bind("dc","http://purl.org/dc/elements/1.1/")
        self._namespaces.bind("rdfs","http://www.w3.org/2000/01/rdf-schema#")
        self._namespaces.bind("dbo","http://dbpedia.org/ontology/")
        self._namespaces.bind("rss","http://purl.org/rss/1.0/")
        self._namespaces.bind("skos","http://www.w3.org/2004/02/skos/core#")
        self._namespaces.bind("void","http://rdfs.org/ns/void#")
        self._namespaces.bind("dn","http://purl.org/datanode/ns/")
        self._namespaces.bind("ex","http://purl.org/datanode/ex/")
        self._namespaces.bind("dct","http://purl.org/dc/terms/")
    
    def entities(self):
        entities = []
        for s,p,o in self._data.triples( (None, None, None) ):
            if s not in entities:
                entities.append(s)
            if o not in entities and isinstance(o, URIRef):
                entities.append(s)
        return entities


        
    def jsobject(self):
        print('var nodes=[')
        first = True
        nodes=[]
        for s,p in list(self._data.subject_predicates(None)):
            if str(p).startswith('http://purl.org/datanode/ns/') and s not in nodes:
                nodes.append(s)
        for p,s in list(self._data.predicate_objects(None)):
            if str(p).startswith('http://purl.org/datanode/ns/') and s not in nodes:
                nodes.append(s)

        for node in nodes:
            if first:
                first = False
            else:
                print(",")
                
            links = []
            for (pred, obj) in self._data.predicate_objects(node):
                if str(pred).startswith('http://purl.org/datanode/ns/'):
                    links.append({
                        'target': nodes.index(obj),
                        'label': self.label(pred),
                        'uri': pred.encode('utf-8'),
                        'type': 'default',
                        'weight': 1
                    })
            
            n = {
                'id': str(nodes.index(node)),
                'uri': node.encode('utf-8'),
                'label': self.label(node),
                'links': links
            }
            print(json.dumps(n))
        print("]")

    def label(self, o, useHtml = False):
        if o in vok.graph().subjects():
            return vok.label(o)
        else:
            if isinstance(o, URIRef):
                if str(o).startswith('http://purl.org/datanode/ex/'):
                    # hack, override ex namespace
                    ns = "http://purl.org/datanode/ex/"
                    if str(o).find('#') > -1:
                        ns = str(o).rsplit('#')[0] + '#'
                    else:
                        ns = str(o).rsplit('/')[0] + '/'
                    self._namespaces.bind('ex', ns)
                try:
                    return next(iter(self._data.objects(o, RDFS.label) or []), self._namespaces.qname(o))
                except:
                    return o.encode('utf-8')
            if isinstance(o, BNode):
                return str(o)
            else:
                # literal
                if useHtml:
                    html = re.sub("([^\.?!]{20,30}[\.?!])|([^,0-9]{20,30},)", "\0<br/>", str(o))
                    html = re.sub("[0-9]{1,3}\)", "<br/>\0", html)
                    html = cgi.escape(html);
                    html = "<html>" + html + "</html>"
                    html = re.sub("\<",'&lt;', html)
                    html = re.sub("\>",'&gt;', html)
                    return html
                else:
                    return o.encode('utf-8')
        

    
    def graphml(self):
            print("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>")
            print("<graphml \n\txmlns=\"http://graphml.graphdrawing.org/xmlns\" ")
            print("\txmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" ")
            print("\txmlns:y=\"http://www.yworks.com/xml/graphml\" ")
            print("\txmlns:yed=\"http://www.yworks.com/xml/yed/3\" ")
            print("\txsi:schemaLocation=\"http://graphml.graphdrawing.org/xmlns http://www.yworks.com/xml/schema/graphml/1.1/ygraphml.xsd\">")
            print("\t\t<key attr.name=\"label\" attr.type=\"string\" for=\"node\" id=\"k_label\"/>")
            print("\t\t<key for=\"node\" id=\"key_nodegraphics\" yfiles.type=\"nodegraphics\"/>")
            print("\t\t<key for=\"edge\" id=\"key_edgegraphics\" yfiles.type=\"edgegraphics\"/>")
            #puts "<key attr.name=\"description\" attr.type=\"string\" for=\"node\" id=\"d5\"/>"
            print("\t<graph edgedefault=\"directed\" id=\"G\">")
            
            for node in self.entities():
                print ("<!-- " + node + " -->")
                nodel = self.label(node, True)
                print ("<!-- " + nodel.encode('utf-8') + " -->")
                print ("\t\t<node id=\"n" + str(id(node)) + "\">")
                print ("\t\t\t<data key=\"k_label\"><![CDATA[" + node.encode('utf-8') + "]]></data>")
                print ("\t\t\t<data key=\"d5\"><![CDATA[" + node.encode('utf-8') + "]]></data>")
                print ("\t\t\t<data key=\"key_nodegraphics\">")
                print ("\t\t\t\t<y:GenericNode configuration=\"com.yworks.flowchart.start1\">")
                print ("\t\t\t\t\t<y:Fill color=\"#FFFFFF\" transparent=\"true\"/>")
                print ("\t\t\t\t\t<y:BorderStyle color=\"#FFFFFF\" transparent=\"true\" type=\"line\" width=\"0.0\"/>")
                print ("\t\t\t\t\t<y:NodeLabel ")
                print (">" + nodel.encode('utf-8') + "<y:LabelModel><y:SmartNodeLabelModel distance=\"4.0\"/></y:LabelModel> ")
                print ("<y:ModelParameter> ")
                print ("<y:SmartNodeLabelModelParameter ")
                print ("/>")
                print ("</y:ModelParameter>")
                print ("</y:NodeLabel>")
                print ("\t\t\t\t</y:GenericNode>")
                print ("\t\t\t</data>")
                print ("\t\t</node>")
                for (link, to) in self._data.predicate_objects(node):
                    edgeId = 0
                    print("<!-- " + to.encode('utf-8') + " -->")
                    edgeId = str(id(node)) + "_" + str(id(link)) + "_" + str(id(to))
                    edgel = self.label(link, True)
                    print("\t\t<edge id=\"e_" + edgeId + "\" source=\"n" + str(id(node)) + "\" target=\"n" + str(id(to)) + "\">")
                    print("\t\t\t<data key=\"key_edgegraphics\">")
                    print("\t\t\t<y:BezierEdge>")
                    print("\t\t\t\t<y:LineStyle color=\"#000000\" type=\"line\" width=\"2.0\"/>")
                    print("\t\t\t\t<y:Arrows source=\"none\" target=\"white_delta\"/>")
                    print("\t\t\t\t<y:EdgeLabel>" + edgel.encode('utf-8') + "</y:EdgeLabel>")
                    print("\t\t\t</y:BezierEdge>")
                    print("\t\t\t</data>")
                    print("\t\t</edge>")

            print("\t</graph>")
            print("</graphml>")

if __name__ == "__main__" :
    parser = argparse.ArgumentParser(description='Rdfdia python script to generate GraphML diagrams from RDF data snippets.', version='1.0.0c - Enrico Daga')
    parser.add_argument('-f', '--file', metavar='F', type=str, nargs=1,
                        help='RDF data file name.')
    parser.add_argument('-V', '--vocabulary', metavar='V', type=str, nargs=1,
                        help='RDF Vocabulary file.')
    parser.add_argument('-o', '--output', metavar='O', type=str, nargs=1,
                        help='Output type: graphml or js.', choices=['graphml','js'])
    parser.add_argument('-m', '--modifier', metavar='V', type=str, nargs=1,
                        help='Modifier (links | all).', choices=['links','all'], default='all')
    args = parser.parse_args()

    if not os.path.isfile(str(args.file[0])):
        print('Error! File ' + args.file[0] + ' does not exists')
        print('')
        parser.print_help()
        exit()
    if not os.path.isfile(str(args.vocabulary[0])):
        print('Error! File ' + args.vocabulary[0] + ' does not exists')
        print('')
        parser.print_help()
        exit()

    data = args.file[0]
    voc = args.vocabulary[0]
    output = args.output[0]
    modifier = args.modifier[0]
    
    vok = Vok(os.path.abspath(args.file[0]), 'http://purl.org/datanode/ns/relatedWith')
    dia = Dia(data, vok, modifier)
    
    if output == 'graphml':
        dia.graphml()

    if output == 'js':
        dia.jsobject()

    #print(args.accumulate(args.integers))
    
